lib/average_daily_balance/main.py

```diff
@@ -0,0 +1,117 @@
+import os
+from datetime import datetime, timedelta
+from typing import Optional
+
+import pandas as pd
+from pngme.api import Client
+
+
+def get_avg_daily_balance(
+    client: Client, user_uuid: str, utc_starttime: datetime, utc_endtime: datetime
+) -> float:
+    """Calculates the average daily account balance for a user across all
+       of their accounts for a given time window. 
+
+       Typical date ranges are last 30 days, 31-60 days and 61-90 days.
+       
+    Args:
+        client: Pngme API client
+        user_uuid: the Pngme user_uuid for the mobile phone user
+        utc_starttime: the datetime for the left-hand-side of the time-window
+        utc_endtime: the datetime for the right-hand-side of the time-window
+    Returns:
+        the average daily account balance for a given time window
+    """
+    institutions = client.institutions.get(user_uuid)
+    institution_balances = []
+
+    today = datetime.today()
+    all_pages_utc_endtime = datetime(today.year, today.month, today.day)
+    all_pages_starttime = all_pages_utc_endtime - timedelta(days=1e5)  # get all pages
+
+    # 0. Get all balances for a user
+    for institution in institutions:
+        institution_id = institution.institution_id
+        balances = client.balances.get(
+            user_uuid=user_uuid,
+            institution_id=institution_id,
+            utc_starttime=all_pages_starttime,
+            utc_endtime=all_pages_utc_endtime,
+        )
+        # convert to records
+        balances = [
+            balance.dict()
+            for balance in balances
+            if (balance.account_type == "depository")
+        ]
+        balances = pd.DataFrame.from_records(balances)
+        balances = balances.assign(institution_name=institution_id)
+        institution_balances.append(balances)
+
+    # 1. Create DF of institution balances
+    institution_balances = pd.concat(institution_balances)
+    if institution_balances.empty:
+        return None
+
+    # 2. Sort and create a column for day, filter by time window
+    institution_balances = institution_balances.sort_values("ts")
+    institution_balances["yyyymmdd"] = institution_balances["ts"].apply(
+        lambda epoch_time: datetime.fromtimestamp(epoch_time)
+    )
+    institution_balances["yyyymmdd"] = institution_balances["yyyymmdd"].dt.round("D")
+
+    # 3. Filter time window
+    institution_balances = institution_balances[
+        (institution_balances["yyyymmdd"] >= utc_starttime)
+        & (institution_balances["yyyymmdd"] <= utc_endtime)
+    ]
+    if institution_balances.empty:
+        return None
+
+    # 4. Get end of day balances, if an account changes balances three times a day ($100, $20, $120), take the last one ($120)
+    eod_balances = institution_balances.groupby(
+        ["yyyymmdd", "account_number", "institution_name"]
+    ).tail(1)
+    eod_balances = eod_balances.set_index("yyyymmdd")
```

Comment:
You can also use `inplace` argument in `set_index` for short

Reply:
not a fan of inplace